src/libsystemd-network/dhcp6-option.c

```diff
@@ -410,14 +404,64 @@ int dhcp6_option_parse(
         return 0;
 }
 
-int dhcp6_option_parse_status(DHCP6Option *option, size_t len) {
-        DHCP6StatusOption *statusopt = (DHCP6StatusOption *)option;
+int dhcp6_option_parse_status(const uint8_t *data, size_t data_len, char **ret_status_message) {
+        assert(data);
 
-        if (len < sizeof(DHCP6StatusOption) ||
-            be16toh(option->len) + offsetof(DHCP6Option, data) < sizeof(DHCP6StatusOption))
-                return -ENOBUFS;
+        if (data_len < sizeof(uint16_t))
+                return -EBADMSG;
+
+        if (ret_status_message) {
+                char *msg;
 
-        return be16toh(statusopt->status);
+                /* The status message MUST NOT be null-terminated. See section 21.13 of RFC8415. */
+                msg = memdup_suffix0(data + sizeof(uint16_t), data_len - sizeof(uint16_t));
+                if (!msg)
+                        return -ENOMEM;
+
+                *ret_status_message = msg;
+        }
+
+        return be16toh(*((const be16_t*) data));
```

Comment:
so i take it we assume the data is always properly aligned? because if there's the risk that it isn't we should maybe use unaligned_read_be16() hre

Reply:
I am not familiar with alignedness. What is alignedness? What happens when it is not aligned?

Reply:
on most cpus 32bit integers must be placed in memory at addresses divisable by 4. Similar 16bit integers, must be placed in addresses divisable by 2, and 64bit integers at addresses divisable by 8. otoh 8bit integers can be placed at any location in memory. Now, thankfully the compiler takes mostly care of this, and will put empty spaces in structs to make sure this "alignment" is held and things just work.

Now, when we read stuff off disk or from network packets the data might not be aligned properly: i.e. you might have a uint8_t at the beginning followed immediately by an uint32_t or so. In that case you cannot just access it as uin32_t. What you have to do is to read it as 4 separate 8bit values, and then merge them with bit shifting. This is what unaligned_read_be32() actually does for you: it reads a big-endian 32bit value regardless where it is located in memory, aigned or not.

Now, well designed network protocols actually arrange their packets so that alignment is in place, but older ones might not, in particular when it comes to 64bit values (simply because older cpus couldn't handle 64bit integers natively anyway, and thus treated them as two 32bit integers, so that alignment didn't matter).

so, one can mark structs as "_packed_", in which case gcc will deal with unaligned fields (but that only works if the beginning of the struct is maximally aligned). or one can use unaligned_read_be32() and related helpers so that things work in any case. Or the network protocol is carefully enough design to always have everything maximally aligned... Not sure dhcp qualifies.

Reply:
if you access an integer with a pointer that is incorrectly aligned then x86 family cpus will still be able to read the value, but be awfully slow doing so. other cpus will abort your program with SIGBUS.

Reply:
also see: https://en.wikipedia.org/wiki/Data_structure_alignment

Reply:
Oh, then it seems we need to use unaligned_read_xxx() at so many places... I will address that. Thank you.

Reply:
Ah, fortunately, not many, but only this place. Nice catch!